File: scripts/ci/auto-repair/openhands-repair-advisor.mjs
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { createHash } from 'node:crypto';

const SHA_RE = /^[a-f0-9]{40}$/u;

function git(targetDir, args) {
  return execFileSync('git', ['-C', targetDir, ...args], { encoding: 'utf8' }).trim();
}

export function buildOpenHandsAdvisorPrompt({ targetSha, failureFingerprint, diagnosis, failureLog }) {
  return [
    'You are the OpenHands repair advisor inside FLIXO.',
    'Your role is advisory repair engineering only.',
    '',
    'HARD BOUNDARIES:',
    '- Work only in this detached temporary workspace.',
    '- Never create or switch branches.',
    '- Never commit, push, merge, mutate refs, or access main.',
    '- Never change tests, CI workflows, security controls, certification gates, or control-plane files.',
    '- Never add dependencies or redesign architecture.',
    '- Diagnose the exact current failure and propose the smallest source-only repair.',
    '- You may run targeted diagnostics/tests in the detached workspace.',
    '- Do not claim GREEN, certification, or successful publication.',
    '- The FLIXO repair engine remains the only mutation/commit authority.',
    '',
    'EXACT IDENTITY:',
    JSON.stringify({
      targetSha,
      failureFingerprint,
      diagnosisSha: diagnosis?.targetSha ?? null,
      failureLogSha256: createHash('sha256').update(String(failureLog ?? ''), 'utf8').digest('hex'),
    }),
    '',
    'CURRENT DIAGNOSIS:',
    JSON.stringify(diagnosis ?? null, null, 2).slice(0, 16000),
    '',
    'TASK:',
    '1. Reproduce or inspect the failure.',
    '2. Identify the most likely causal source file and invariant violation.',
    '3. Apply the minimal candidate source patch in this detached workspace.',
    '4. Run the narrowest useful verification.',
    '5. Stop and leave the candidate patch uncommitted.',
    '6. Do not edit tests or infrastructure merely to make the failure disappear.',
  ].join('\n');
}

export function runOpenHandsRepairAdvisor({
  targetDir = process.cwd(),
  targetSha,
  failureFingerprint,
  failureLog = '',
  diagnosis = null,
} = {}) {
  const precomputedPath = process.env.FLIXO_OPENHANDS_PRIORITY_PATH ?? '';
  if (precomputedPath && fs.existsSync(precomputedPath)) {
    try {
      const precomputed = JSON.parse(fs.readFileSync(precomputedPath, 'utf8'));
      if (precomputed?.targetSha === targetSha && precomputed?.protocol === 'FLIXO-OPENHANDS-REPAIR-ADVISOR-v1') {
        return Object.freeze({ ...precomputed, reusedPrecomputed: true });
      }
    } catch {
      // Invalid precomputed evidence is ignored; the exact-SHA adapter will re-run and fail closed if unavailable.
    }
  }
  const enabled = process.env.FLIXO_OPENHANDS_ADVISOR_ENABLED === 'true';
  if (!enabled) {
    return Object.freeze({
      status: 'DISABLED',
      protocol: 'FLIXO-OPENHANDS-REPAIR-ADVISOR-v1',
      targetSha: targetSha ?? null,
      mutationAuthority: 'NONE',
    });
  }

  const currentSha = git(targetDir, ['rev-parse', 'HEAD']);
  const branch = git(targetDir, ['branch', '--show-current']);
  if (!SHA_RE.test(String(targetSha ?? '')) || currentSha !== targetSha) {
    return Object.freeze({ status: 'BLOCKED', reason: 'OPENHANDS_TARGET_SHA_MISMATCH', currentSha, targetSha });
  }
  if (branch !== 'execution') {
    return Object.freeze({ status: 'BLOCKED', reason: 'OPENHANDS_MUTATION_BRANCH_INVALID', branch });
  }
  if (git(targetDir, ['status', '--porcelain'])) {
    return Object.freeze({ status: 'BLOCKED', reason: 'OPENHANDS_SOURCE_WORKTREE_DIRTY' });
  }

  const tempRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'flixo-openhands-'));
  const workspace = path.join(tempRoot, 'workspace');
  const failurePath = path.join(tempRoot, 'failure.log');
  const promptPath = path.join(tempRoot, 'prompt.txt');
  const outputPath = path.join(tempRoot, 'advisor.json');

  try {
    execFileSync('git', ['-C', targetDir, 'worktree', 'add', '--detach', workspace, targetSha], {
      stdio: 'pipe',
    });
    fs.writeFileSync(failurePath, String(failureLog ?? ''), 'utf8');
    fs.writeFileSync(
      promptPath,
      buildOpenHandsAdvisorPrompt({ targetSha, failureFingerprint, diagnosis, failureLog }),
      'utf8',
    );

    const python = process.env.FLIXO_OPENHANDS_PYTHON || 'python3';
    const env = {
      ...process.env,
      FLIXO_OPENHANDS_WORKSPACE: workspace,
      FLIXO_OPENHANDS_TARGET_SHA: targetSha,
      FLIXO_OPENHANDS_FAILURE_LOG: failurePath,
      FLIXO_OPENHANDS_PROMPT_PATH: promptPath,
      FLIXO_OPENHANDS_OUTPUT: outputPath,
    };

    let raw = '';
    try {
      raw = execFileSync(python, ['scripts/ci/auto-repair/openhands-repair-advisor.py'], {
        cwd: targetDir,
        env,
        encoding: 'utf8',
        timeout: Number(process.env.FLIXO_OPENHANDS_TIMEOUT_MS ?? 240000),
        maxBuffer: 8 * 1024 * 1024,
      });
    } catch (error) {
      if (fs.existsSync(outputPath)) {
        return JSON.parse(fs.readFileSync(outputPath, 'utf8'));
      }
      return Object.freeze({
        status: 'BLOCKED_EXTERNAL',
        protocol: 'FLIXO-OPENHANDS-REPAIR-ADVISOR-v1',
        reason: 'OPENHANDS_RUNTIME_UNAVAILABLE',
        errorType: error?.name ?? 'Error',
        error: String(error?.message ?? error).slice(0, 1000),
        targetSha,
      });
    }

    if (fs.existsSync(outputPath)) {
      return JSON.parse(fs.readFileSync(outputPath, 'utf8'));
    }

    try {
      return JSON.parse(raw.trim().split('\n').at(-1) ?? '{}');
    } catch {
      return Object.freeze({
        status: 'BLOCKED_EXTERNAL',
        protocol: 'FLIXO-OPENHANDS-REPAIR-ADVISOR-v1',
        reason: 'OPENHANDS_INVALID_OUTPUT',
        targetSha,
      });
    }
  } finally {
    try {
      execFileSync('git', ['-C', targetDir, 'worktree', 'remove', '--force', workspace], { stdio: 'ignore' });
    } catch {
      // Preserve the primary advisor evidence; cleanup failure is external and must not mutate source.
    }
    fs.rmSync(tempRoot, { recursive: true, force: true });
  }
}

if (process.argv[1]?.endsWith('openhands-repair-advisor.mjs') && process.argv[2] === '--self-test') {
  process.env.FLIXO_OPENHANDS_ADVISOR_ENABLED = 'false';
  const result = runOpenHandsRepairAdvisor({ targetSha: 'a'.repeat(40) });
  if (result.status !== 'DISABLED') process.exit(1);
  console.log('OPENHANDS_REPAIR_ADVISOR_SELF_TEST=PASS');
}

File: scripts/ci/auto-repair/openhands-repair-advisor.py
#!/usr/bin/env python3
"""OpenHands advisory repair worker.

This worker MUST run only inside a detached temporary worktree created by the
Node adapter. It may inspect, edit, and test that temporary copy, but it never
owns the FLIXO mutation/commit lane.
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path
from typing import Any

from pydantic import SecretStr
from openhands.sdk import Agent, Conversation, Event, LLM, LLMConvertibleEvent, Tool
from openhands.tools.file_editor import FileEditorTool
from openhands.tools.task_tracker import TaskTrackerTool
from openhands.tools.terminal import TerminalTool


def git(cwd: Path, *args: str) -> str:
    result = subprocess.run(
        ["git", "-C", str(cwd), *args],
        check=True,
        text=True,
        capture_output=True,
    )
    return result.stdout.strip()


def emit(payload: dict[str, Any]) -> int:
    output = os.environ.get("FLIXO_OPENHANDS_OUTPUT", "/tmp/flixo-openhands-advisor.json")
    Path(output).write_text(json.dumps(payload, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(json.dumps(payload, ensure_ascii=False))
    return 0


def main() -> int:
    workspace = Path(os.environ["FLIXO_OPENHANDS_WORKSPACE"]).resolve()
    target_sha = os.environ["FLIXO_OPENHANDS_TARGET_SHA"]
    failure_log = os.environ["FLIXO_OPENHANDS_FAILURE_LOG"]
    prompt_path = os.environ["FLIXO_OPENHANDS_PROMPT_PATH"]

    if not workspace.is_dir():
        return emit({"status": "BLOCKED", "reason": "OPENHANDS_WORKSPACE_MISSING"})
    if git(workspace, "rev-parse", "HEAD") != target_sha:
        return emit({"status": "BLOCKED", "reason": "OPENHANDS_TARGET_SHA_MISMATCH"})
    if git(workspace, "branch", "--show-current") not in ("", "(HEAD detached at " + target_sha + ")"):
        return emit({"status": "BLOCKED", "reason": "OPENHANDS_WORKSPACE_NOT_DETACHED"})

    api_key = (
        os.environ.get("FLIXO_OPENHANDS_API_KEY")
        or os.environ.get("OPENAI_API_KEY")
        or os.environ.get("OPENROUTER_API_KEY")
    )
    if not api_key:
        return emit({"status": "BLOCKED_EXTERNAL", "reason": "OPENHANDS_LLM_CREDENTIAL_MISSING"})

    model = (
        os.environ.get("FLIXO_OPENHANDS_MODEL")
        or (os.environ.get("OPENROUTER_MODEL") if os.environ.get("OPENROUTER_API_KEY") else None)
        or os.environ.get("OPENAI_MODEL")
        or "gpt-5.5"
    )
    base_url = os.environ.get("FLIXO_OPENHANDS_BASE_URL")
    if not base_url and os.environ.get("OPENROUTER_API_KEY"):
        base_url = "https://openrouter.ai/api/v1"

    prompt = Path(prompt_path).read_text(encoding="utf-8")
    failure_tail = Path(failure_log).read_text(encoding="utf-8")[-24000:]

    llm = LLM(
        usage_id="flixo-repair-advisor",
        model=model,
        base_url=base_url,
        api_key=SecretStr(api_key),
    )
    agent = Agent(
        llm=llm,
        tools=[
            Tool(name=TerminalTool.name),
            Tool(name=FileEditorTool.name),
            Tool(name=TaskTrackerTool.name),
        ],
    )

    events: list[str] = []

    def callback(event: Event) -> None:
        if isinstance(event, LLMConvertibleEvent):
            events.append(str(event.to_llm_message()))

    task = (
        prompt
        + "\n\nFAILURE LOG TAIL:\n"
        + failure_tail
    )
    conversation = Conversation(agent=agent, workspace=workspace, callbacks=[callback])
    conversation.send_message(task)
    conversation.run()

    changed = git(workspace, "diff", "--name-only")
    diff = subprocess.run(
        ["git", "-C", str(workspace), "diff", "--binary"],
        check=True,
        text=True,
        capture_output=True,
    ).stdout
    summary = git(workspace, "diff", "--stat")

    return emit(
        {
            "schemaVersion": 1,
            "protocol": "FLIXO-OPENHANDS-REPAIR-ADVISOR-v1",
            "status": "COMPLETED",
            "targetSha": target_sha,
            "changedFiles": [line for line in changed.splitlines() if line],
            "diffStat": summary,
            "candidatePatch": diff[:1_000_000],
            "candidatePatchTruncated": len(diff) > 1_000_000,
            "llmEventCount": len(events),
            "lastEvents": events[-8:],
            "mutationAuthority": "NONE",
            "commitAuthority": "NONE",
            "mainMutation": False,
            "branchCreation": False,
            "verificationAuthority": "FLIXO_CANONICAL_GATES",
        }
    )


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        raise SystemExit(
            emit(
                {
                    "status": "BLOCKED_EXTERNAL",
                    "reason": "OPENHANDS_RUNTIME_FAILURE",
                    "errorType": type(exc).__name__,
                    "error": str(exc)[:1000],
                }
            )
        )
